### PR TITLE
Create a HttpContext Based Managed Tracer

### LINKED ISSUE #835: Consider accessing IManagedTracers from the IHttpContextAccessor
Currently for Google.Cloud.Diagnostics.AspNetCore access for IManagedTracers is done by dependency injects and each IManagedTraceris scoped to the current request.  This requires users to have to inject the scope based IManagedTracers.

We should consider passing around instead a singleton Func<IManagedTracer> that relies on IHttpContextAccessor where the user would not have to worry about scoping.

cc: @jskeet

### PR DESCRIPTION
**Pull request by iantalarico**
This PR Creates an IManagedTracer that is a wrapper around another IManagedTracer which it accesses via the current HttpContext.  This allows for a singleton IManagedTracer.

The IManagedTracer is now created in the middleware right before it is used and then set on the HttpContext. The current HttpContext can always be accessed by a singleton IHttpContextAccessor this allows for a singleton IManagedTracer so users do not have to deal with scoping issues of the IManagedTracer.

This change does mean users who try to trace anything before the trace middleware is registered will just be using a null managed tracer and not tracing anything.  I think this is acceptable for now as this seems like it will be a small use case and we would need to be careful about how we handle the spans.

Fixes #835 

This is a similar access model to the AspNet library.

### CODE REVIEW

**Comment by evildour** [comment]
There are test failures. Should I hold off on reviewing until they are fixed or is it safe to review?

**Comment by iantalarico** [comment]
It's safe to review.  I just submitted a fix. Thanks Mike!

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceMiddleware.cs [hunk 1]
```diff
@@ -29,24 +29,36 @@ namespace Google.Cloud.Diagnostics.AspNetCore
...
         }
 
         /// <summary>
-        /// Invokes the next <see cref="RequestDelegate"/> and trace the time 
+        /// Invokes the next <see cref="RequestDelegate"/> and traces the time 
         /// taken for the next delegate to run, reporting the results to the
         /// Stackdriver Trace API.
         /// </summary>
-        public async Task Invoke(HttpContext httpContext, IManagedTracer tracer)
+        public async Task Invoke(HttpContext httpContext, TraceHeaderContext traceHeaderContext)
```
Comment (evildour) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceMiddleware.cs]: Might be good to add a \<param\> tag for this, which also indicates that it cannot be null.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceMiddleware.cs [hunk 2]
```diff
@@ -29,24 +29,36 @@ namespace Google.Cloud.Diagnostics.AspNetCore
...
+        private readonly IManagedTracerFactory _tracerFactory;
+        private readonly IHttpContextAccessor _accessor;
 
         /// <summary>
         /// Create a new instance of <see cref="CloudTraceMiddleware"/>.
         /// </summary>
         /// <param name="next">The next request delegate. Cannot be null.</param>
-        public CloudTraceMiddleware(RequestDelegate next)
+        /// <param name="tracerFactory">A factory to create <see cref="IManagedTracer"/>s.</param>
+        /// <param name="accessor">Access to <see cref="HttpContext"/>.</param>
```
Comment (evildour) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceMiddleware.cs]: Indicate that these two cannot be null.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextManagedTracer.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Google.Cloud.Diagnostics.Common;
+using Microsoft.AspNetCore.Http;
+using Google.Api.Gax;
+
+namespace Google.Cloud.Diagnostics.AspNetCore
+{
+    /// <summary>
+    /// An <see cref="IManagedTracer"/> that will pull the the current tracer from
```
Comment (evildour) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextManagedTracer.cs]: Nit: double "the" here

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextTracerManager.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        public static void SetCurrentTracer(IHttpContextAccessor accessor, IManagedTracer tracer)
+        {
+            if (accessor?.HttpContext?.Items != null)
+            {
+                accessor.HttpContext.Items[TraceKey] = tracer;
+            }
+        }
+
+        /// <summary>
+        /// Gets the current <see cref="IManagedTracer"/> or a <see cref="NullManagedTracer"/> if non exists.
```
Comment (evildour) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextTracerManager.cs]: Nit: "non" => "none"

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs [hunk 5]
```diff
@@ -138,31 +135,24 @@ public static void UseGoogleTrace(this IApplicationBuilder app)
             var tracerFactory = new ManagedTracerFactory(projectId, consumer,
                 RateLimitingTraceOptionsFactory.Create(config), traceIdFactory);
 
+            services.AddScoped(CreateTraceHeaderContext);
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs]: So this isn't used by anything else, it's just available if you *want* to inject it in a scoped dependency? If we didn't have this line (and the CreateTraceHeaderContext method) would anything else fail?

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextManagedTracer.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    /// the <see cref="HttpContext"/>. The current tracer is set per request in the
+    /// <see cref="CloudTraceMiddleware"/> via the <see cref="ContextTracerManager"/>.
+    /// This allows for this to be a singleton <see cref="IManagedTracer"/> so users
+    /// do not need to worry about scoping of the tracer.
+    /// </summary>
+    internal sealed class ContextManagedTracer : IManagedTracer
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextManagedTracer.cs]: If we find we want a similar class elsewhere, we could have a DelegatingTracer which accepted a `Func<IManagedTracer>`, and made each call just delegate to that and then pass the arguments to it, abstracting the `ContextTracerManager.GetCurrentTracer(_accessor)` part.

Do we already have anything like this in the AspNet code?

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextTracerManager.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    /// <summary>
+    /// Manages access to the current <see cref="IManagedTracer"/>.
+    /// </summary>
+    internal static class ContextTracerManager
+    {
+        ///<summary>The key to save the tracer under in the <see cref="HttpContext"/>.</summary> 
+        internal const string TraceKey = "Google.Cloud.Diagnostics.AspNetCore.Trace";
+
+        /// <summary>
+        /// Sets the current <see cref="IManagedTracer"/>.
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextTracerManager.cs]: And this is called by the middleware on each request, yes? Maybe add a comment about that.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs (hunk 5 above)
Comment (iantalarico) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs]: This is used in the middleware to create the managed tracer.  We could take that logic and move it here too.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceMiddleware.cs (hunk 2 above)
Comment (iantalarico) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceMiddleware.cs]: Done

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceMiddleware.cs (hunk 1 above)
Comment (iantalarico) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceMiddleware.cs]: Done

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextManagedTracer.cs (hunk 3 above)
Comment (iantalarico) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextManagedTracer.cs]: done

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextManagedTracer.cs (hunk 6 above)
Comment (iantalarico) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextManagedTracer.cs]: We don't have anything yet but I need to apply this general change to the ASP.NET code and will make this change then.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextTracerManager.cs (hunk 7 above)
Comment (iantalarico) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextTracerManager.cs]: Done

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextTracerManager.cs (hunk 4 above)
Comment (iantalarico) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/ContextTracerManager.cs]: Done

**Comment by jskeet** [comment]
Ah - so `public async Task Invoke(HttpContext httpContext, IManagedTracer tracer)` in `CloudTraceMiddleware` gets the scoped tracer injected automatically? That makes sense. Presumably we *could* use the context-based one there too, but we don't need to.

Okay, LGTM (IIRC - it's hard to keep track of where I was on this).

**Pull request merged**